tests/phpunit/includes/testcase-canonical.php

```diff
@@ -1,54 +1,86 @@
+	use PLL_UnitTestCase_Trait;
 
-		if ( ! isset( $expected['url'] ) && ! isset( $expected['qv'] ) ) {
-			$this->markTestSkipped( 'No valid expected output was provided' );
-		}
+	private $options;
 
-		$this->go_to( home_url( $test_url ) );
+	public function setUp() {
+		parent::setUp();
 
-		// Does the redirect match what's expected?
-		$can_url = self::$polylang->filters_links->check_canonical_url( home_url( $test_url ), false ); // FIXME TODO define links ( need $curlang )
-		if ( $wp_can_url = redirect_canonical( $can_url, false ) ) {
-			$parsed_can_url = wp_parse_url( $wp_can_url );
-		} else {
-			$parsed_can_url = wp_parse_url( $can_url );
-		}
+		add_filter( 'wp_using_themes', '__return_true' ); // To pass the test in PLL_Choose_Lang::init() by default.
+		add_filter( 'wp_doing_ajax', '__return_false' );
 
-		// Just test the Path and Query if present
-		if ( isset( $expected['url'] ) ) {
-			$this->assertEquals( $expected['url'], $parsed_can_url['path'] . ( ! empty( $parsed_can_url['query'] ) ? '?' . $parsed_can_url['query'] : '' ) );
-		}
+		$this->options = array_merge(
+			PLL_Install::get_default_options(),
+			array(
+				'default_lang' => 'en',
+				'hide_default' => 0,
+				'post_types'   => array(
+					'cpt' => 'pllcanonical',
+					// translate the cpt // FIXME /!\ 'after_setup_theme' already fired and the list of translated post types is already cached :(
+				),
+			)
+		);
+	}
 
-		if ( ! isset( $expected['qv'] ) ) {
-			return;
+	/**
+	 * Set up the Polylang environment before testing canonical redirects.
+	 *
+	 * @param string $test_url
+	 * @param string $expected
+	 * @param int    $ticket
+	 * @param array  $expected_doing_it_wrong
+	 */
+	public function assertCanonical( $test_url, $expected, $ticket = 0, $expected_doing_it_wrong = array() ) {
+		global $wp_rewrite;
+
+		if ( did_action( 'pll_language_defined' ) ) {
+			$this->fail( 'Canonical tests MUST have only one call to PLL_UnitTestCase_Canonical::assertCanonical() per test.' );
 		}
 
-		// "make" that the request and check the query is correct
-		$this->go_to( $can_url );
+		// Needed by {@see pll_requested_url()}.
+		$_SERVER['REQUEST_URI'] = $test_url;
+
+		$model = new PLL_Model( $this->options );
+
+		// switch to pretty permalinks
+		$wp_rewrite->init();
+		$wp_rewrite->set_permalink_structure( $this->structure );
 
-		// Are all query vars accounted for, And correct?
-		global $wp;
+		// register post types and taxonomies
+		$model->post->register_taxonomy(); // needs this for 'lang' query var
+		create_initial_taxonomies();
 
-		$query_vars = array_diff( $wp->query_vars, $wp->extra_query_vars );
-		if ( ! empty( $parsed_can_url['query'] ) ) {
-			parse_str( $parsed_can_url['query'], $_qv );
+		// reset the links model according to the permalink structure
+		$links_model    = $model->get_links_model();
+		self::$polylang = new PLL_Frontend( $links_model );
+		self::$polylang->init();
+		do_action_ref_array( 'pll_init', array( self::$polylang ) );
```

Comment:
This is looking more and more like `Polylang::init()`. What would be the issues if I wanted to replace the manual intialization by a call to `Polylang::init()`?